network/src/application/storage.rs

```diff
@@ -1,18 +1,150 @@
+            .into_iter()
+            .map(|peer_id| PeerNetworkId::new(network_id, peer_id))
+            .collect()
+    }
+
+    /// Read a clone of the entire state
+    pub fn read_all(&self, network_id: NetworkId) -> HashMap<PeerNetworkId, PeerInfo> {
+        to_peer_network_ids(network_id, self.get_network(network_id).read_all())
+    }
+
+    /// Insert new entry
+    pub fn insert(&self, peer_network_id: PeerNetworkId, new_value: PeerInfo) {
+        self.get_network(peer_network_id.network_id())
+            .insert(peer_network_id.peer_id(), new_value)
+    }
+
+    /// Remove old entries
+    pub fn remove(&self, peer_network_id: &PeerNetworkId) {
+        self.get_network(peer_network_id.network_id())
+            .remove(&peer_network_id.peer_id())
+    }
+
+    /// Take in a function to modify the data, must handle concurrency control with the input function
+    pub fn write<F: FnOnce(&mut Entry<PeerId, PeerInfo>) -> Result<(), PeerError>>(
+        &self,
+        peer_network_id: PeerNetworkId,
+        modifier: F,
+    ) -> Result<(), PeerError> {
+        self.get_network(peer_network_id.network_id())
+            .write(peer_network_id.peer_id(), modifier)
+    }
+
+    /// Get the underlying `RwLock` of the map.  Usage is discouraged as it leads to the possiblity of
+    /// leaving the lock held for a long period of time.  However, not everything fits into the `write`
+    /// model.
+    pub fn write_lock(
+        &self,
+        network_id: NetworkId,
+    ) -> RwLockWriteGuard<'_, HashMap<PeerId, PeerInfo>> {
+        self.get_network(network_id).write_lock()
+    }
+
+    pub fn insert_connection(
+        &self,
+        network_id: NetworkId,
+        connection_metadata: ConnectionMetadata,
+    ) {
+        self.write_lock(network_id)
+            .entry(connection_metadata.remote_peer_id)
+            .and_modify(|entry| entry.active_connection = connection_metadata.clone())
+            .or_insert_with(|| PeerInfo::new(connection_metadata));
+    }
+
+    pub fn remove_connection(
+        &self,
+        network_id: NetworkId,
+        connection_metadata: &ConnectionMetadata,
+    ) {
+        let mut map = self.write_lock(network_id);
+
+        // Don't remove the peer if the connection doesn't match!
+        if let Entry::Occupied(entry) = map.entry(connection_metadata.remote_peer_id) {
+            // For now, remove the peer entirely, we could in the future have multiple connections for a peer
+            if entry.get().active_connection.connection_id == connection_metadata.connection_id {
+                entry.remove();
+            }
+        }
+    }
+}
+
+fn to_peer_network_ids(
+    network_id: NetworkId,
+    map: HashMap<PeerId, PeerInfo>,
+) -> HashMap<PeerNetworkId, PeerInfo> {
+    map.into_iter()
+        .map(|(peer_id, peer_info)| (PeerNetworkId::new(network_id, peer_id), peer_info))
+        .collect()
+}
 
 /// A generic locking hash map with ability to read before write consistency
```

Comment:
nit: what do you mean by "read before write" consistency? we're using a lock, so everything is serialized : p

Reply:
Basically, I'm trying to hide the locking mechanism from the outside code to prevent locking on excessively long waits etc.

When I mean read before write, you have to take the write lock, do some read, then do a write.  Rather than just write something without what was written before it.  Otherwise, if you did say `storage.read()` followed by an `storage.write()` you might not have what you originally read still be there.

e.g. I have a function that adds `1`.  I read the value `1`, and decide I want to write `2`.  Meanwhile, another thread does the same thing, and updates the value to `10`.  Would the value I want to be is `2` or `11`?

Reply:
i don't know if i would try to hide the fact that it's used concurrently, since the concurrency affects how you can use it. rather just document that it's like a `RwLock<HashMap<_, _>>`, which should give anyone reading an accurate mental model

Reply:
Yeah, if you see here I still failed on my mission to fit everything into `write` as I still needed to create a `write_lock` :(